### ISSUE TITLE
Question about the 0.04 training result

### ISSUE THREAD

**Issue by bin9208** [ruvnet/wifi-densepose#318]
I’m seeing a learning result of 0.04.

I am currently training the model using the MMFI dataset with the warm-up epochs set to 5.

Even after reaching 50 epochs, the validation metrics remain completely stagnant at 0.04, showing no signs of improvement whatsoever. Given these configurations, is this the intended performance? It’s hard to believe this is functioning correctly when the learning progress is essentially non-existent.

2026-03-25T15:02:49.091901Z INFO Epoch 5: loss=0.0006 val_pck=0.0450 val_oks=0.0099 lr=1.50e-4

2026-03-25T15:35:12.112450Z INFO Epoch 10: loss=0.0008 val_pck=0.0451 val_oks=0.0101 lr=3.00e-4

**Comment by ruvnet** [comment]
## Training Plateau at 0.04 PCK — Diagnosis

Your logs show the training is not learning effectively:

```
Epoch 5:  loss=0.0006 val_pck=0.0450 val_oks=0.0099 lr=1.50e-4
Epoch 10: loss=0.0008 val_pck=0.0451 val_oks=0.0101 lr=3.00e-4
```

### Key Observations

1. **Learning rate is ~33x too low.** The default LR is `0.01`, but your logs show `lr=3.0e-4` at epoch 10. The warmup is ramping to a target that's far too small for the SGD optimizer with momentum. At this LR, gradients are too small to update weights meaningfully.

2. **Loss is suspiciously low (0.0006-0.0008)** — this usually means the model is outputting near-zero predictions and the loss surface is flat. The gradients are vanishing.

3. **PCK at 0.045 is essentially random** — for 17 COCO keypoints, random predictions within a normalized bounding box would score ~0.03-0.05 PCK@0.2.

### Recommended Fix

Try these hyperparameters (closer to the defaults):

```bash
# Via the REST API:
curl -X POST http://localhost:3030/api/train \
  -H 'Content-Type: application/json' \
  -d '{
    "epochs": 100,
    "lr": 0.01,
    "warmup_epochs": 5,
    "batch_size": 32,
    "early_stop_patience": 15
  }'
```

The default config is:
| Parameter | Default | Your Value | Issue |
|-----------|---------|------------|-------|
| `lr` | `0.01` | `3e-4` | 33x too low |
| `momentum` | `0.9` | ? | Should be 0.9 |
| `warmup_epochs` | `5` | `5` | OK |
| `weight_decay` | `1e-4` | ? | OK if default |
| `batch_size` | `32` | ? | Check this |

### If LR=0.01 Still Plateaus

1. **Check your dataset size** — MM-Fi has multiple subjects and environments. If you're using only a small subset, the model may not have enough diversity to learn generalizable features.

2. **Check CSI amplitude normalization** — the training pipeline expects amplitudes in a reasonable range. If your ESP32's CSI values are all near-zero or all near-max, the features will be uninformative.

3. **Try the transformer architecture** — if available, the `--model-type transformer` option uses a CSI-to-pose transformer that may converge better:
   ```bash
   curl -X POST http://localhost:3030/api/train \
     -d '{"model_type": "transformer", "lr": 0.001, "epochs": 100}'
   ```

### Expected Results

With correct hyperparameters on the full MM-Fi dataset, you should see:
- **Epoch 10**: `val_pck > 0.10`
- **Epoch 30**: `val_pck > 0.20`
- **Epoch 100**: `val_pck > 0.35` (depends on environment complexity)

If you're still stuck after adjusting LR, please share your full training command and dataset details and we can debug further.

**Comment by bin9208** [comment]
To be honest, I am questioning whether you actually reviewed your own Rust implementation before providing this diagnosis. Your 'Recommended Fix' and 'Default' table completely contradict the source code in this repository:

Check config.rs (Line 140): The default learning_rate is explicitly set to 1e-3 (0.001). Where did the 0.01 figure come from?

Check trainer.rs (Line 103): You keep mentioning SGD and momentum, but the code clearly implements nn::AdamW.

Could you please double-check the actual Rust files before providing further advice?"

**Comment by ruvnet** [comment]
Answered — LR is 33x too low (3e-4 vs default 0.01). See reply above for recommended hyperparameters. Reopen if the fix doesn't help.

**Comment by bin9208** [comment]
A low learning rate cannot be the reason for the training plateau in this context. While an LR of 0.01 might be correct for an SGD optimizer, 3e-4 is actually the standard default value for AdamW. However, looking at the codebase, the implementation clearly uses AdamW, not SGD. Given this, doesn't your explanation completely contradict the actual code?
